R/do_adjustRtime-functions.R

```diff
@@ -582,3 +582,283 @@ adjustRtimeSubset <- function(rtraw, rtadj, subset,
+    nna <- !is.na(idx)
+    idx <- cbind(obs = which(nna), ref = idx[nna])
+    dups <- idx[duplicated(idx[, 2L]), 2L]
+    idx <- idx[!idx[, 2L] %in% dups, , drop = FALSE]
+    data.frame(ref = ref_anchors[idx[, 2L], 2L],
+               obs = obs_peaks[idx[, 1L], 2L])
+}
+
+#' @description
+#'
+#' Compute a model representing the relationship between observed and reference
+#' retention times (`rt_map`) and adjust the raw retention times (`rt_raw`)
+#' based on this.
+#'
+#' @param rt_map `data.frame` with *reference* retention times of LaMas and
+#'     *observed* retention times of matching peaks in the same sample from
+#'     which the retention times in `rt_raw` are.
+#'
+#' @author Carl Brunius
+#'
+#' @noRd
+.adjust_rt_model <- function(rt_raw,
+                             method = c("loess", "gam"),
+                             rt_map,
+                             span = 0.5,
+                             resid_ratio = 3,
+                             zero_weight = 10,
+                             bs = "tp") {
+    model <- .rt_model(method = method,
+                       rt_map, span = span,
+                       resid_ratio = resid_ratio,
+                       zero_weight = zero_weight,
+                       bs = bs)
+    adj <- predict(model, newdata = data.frame(obs = rt_raw))
+    if (is.unsorted(adj, na.rm = TRUE)){
+        warning("Adjusted retention times are not sorted, linear ",
+        "interpolation will be performed for the unsorted data points")
+        adj <- .force_sorted(adj)
+        }
+    idx <- which(rt_raw < min(rt_map$obs))
+    lidx <- length(idx)
+    if (lidx)
+        adj[idx] <- rt_raw[idx] - (rt_raw[lidx + 1L] - adj[lidx + 1L])
+    idx <- which(rt_raw > max(rt_map$obs))
+    if (length(idx))
+        adj[idx] <- rt_raw[idx] - (rt_raw[idx[1L] - 1L] - adj[idx[1L] - 1L])
+    adj
+}
+
+#' @description
+#'
+#' Get a model representing the differences between observed and reference
+#' retention times (parameter `rt_map`). After an initial fit, the model is
+#' re-fitted excluding potential outliers.
+#'
+#' @param rt_map `data.frame` with the observed (column `"obs"`) and reference
+#'     (column `"ref"`) retention time pairs.
+#'
+#' @importFrom stats loess predict resid
+#'
+#' @author Carl Brunius, Philippine Louail
+#'
+#' @noRd
+.rt_model <- function(method = c("loess", "gam"),
+                      rt_map, span = 0.5,
+                      resid_ratio = 3,
+                      zero_weight = 10,
+                      bs = "tp", warnings = FALSE){
+    rt_map <- rt_map[order(rt_map$obs), ]
+    # add first row of c(0,0) to set a fix timepoint.
+    rt_map <- rbind(c(0,0), rt_map)
+    weights <- rep(1, nrow(rt_map))
+    weights[1L] <- zero_weight
+
+    if (method == "gam") {
+        .check_gam_library()
+        model <- mgcv::gam(ref ~ s(obs, bs = bs), weights = weights,
+                           data = rt_map)
+    } else {
+        cw <- tryCatch(
```

Comment:
not sure about the catching of the warnings anymore - the code looks rather clumsy now. what if we keep the code simple and clean and mention in the documentation that, if there are too few data points, not well spaced, loess and gam might throw warnings? and that in that case users might want to reconsider/re-evaluate their lamas and matching to chrom peaks?

Reply:
I would agree with that too i'm not very happy with how it looks. Especially if we prevent model to be build with less than 10 rows in the rtMap. Do you agree with that btw (see above) ?

Reply:
then let's remove the suppression of warnings.

and yes, I'm OK for now with the hard cut of 10 - but keep the message there. would be good to hear Carl's thoughts on that.